### PR TITLE
Add HA configuration

### LINKED ISSUE #7: Make use of high-availability mode

### PR DESCRIPTION
**Pull request by skonto**
Fixes #7, relates to https://github.com/mesosphere/universe/pull/1163

The mesosphere image needs to be built and pushed. Don't have access.
Assumes hdfs is there for the default case (any proper setup should have HA enabled by default) and the storage dir (hdfs://hdfs/flink/recovery) to have been created.
In order to create the dir you need to download the appropriate hdfs distro within your dcos cluster and
execute: 
`dcos hdfs endpoints core-site.xml`
`dcos hdfs endpoints hdfs-site.xml`

to download the required files, to override the defaults within the distro so that finally you can issue commands like:
`hadoop fs -mkdir -p ...`

In order to test it after you build the image under your own account, launch flink with HA, use the following command within a docker container of the flink docker image:

`./bin/flink run -z /default ./examples/batch/WordCount.jar --input /etc/resolv.conf `

Also you need to add to the flink-conf.yaml under the /flink-1.2.0 folder within the container the following settings:
 high-availability: zookeeper
 high-availability.zookeeper.quorum: master.mesos:2181
 high-availability.zookeeper.storageDir: hdfs://hdfs/flink/recovery

### CODE REVIEW

File: container/appmaster/runit/service/flink/run [hunk 1]
```diff
@@ -73,13 +73,37 @@ function add_kerberos_configurations() {
...
+  if [[ "${FLINK_HA_ENABLED}" == true ]]; then
+    export FLINK_JAVA_OPTS="$FLINK_JAVA_OPTS -Dhigh-availability=ZOOKEEPER"
+    add_if_non_empty high-availability.cluster-id "${FLINK_HIGH_AVAILABILITY_CLUSTER_ID}"
+    add_if_non_empty high-availability.storageDir "${FLINK_HIGH_AVAILABILITY_STORAGE_DIR}"
+    add_if_non_empty high-availability.zookeeper.quorum "${FLINK_HIGH_AVAILABILITY_ZOOKEEPER_QUORUM}"
+    add_if_non_empty high-availability.jobmanager.port "${FLINK_HIGH_AVAILABILITY_JOBMANAGER_PORT}"
+    add_if_non_empty high-availability.job.delay "${FLINK_HIGH_AVAILABILITY_JOB_DELAY}"
+    add_if_non_empty high-availability.zookeeper.path.root "${FLINK_HIGH_AVAILABILITY_ZOOKEEPER_PATH_ROOT}"
+    # used for the client at ha services initiated by the mesos app master runner
+    add_if_non_empty ghigh-availability.zookeeper.client.session-timeout \
```
Comment (EronWright) [inline container/appmaster/runit/service/flink/run]: (typo) `ghigh-availability`

File: container/appmaster/runit/service/flink/run [hunk 2]
```diff
@@ -73,13 +73,37 @@ function add_kerberos_configurations() {
   fi
 }
 
+function add_ha_configurations() {
+  if [[ "${FLINK_HA_ENABLED}" == true ]]; then
+    export FLINK_JAVA_OPTS="$FLINK_JAVA_OPTS -Dhigh-availability=ZOOKEEPER"
```
Comment (EronWright) [inline container/appmaster/runit/service/flink/run]: Why did you need to set `high-availability` via `FLINK_JAVA_OPTS` as opposed to the config file?

**Review by EronWright — CHANGES_REQUESTED** [review]

File: container/appmaster/runit/service/flink/run (hunk 2 above)
Comment (skonto) [inline container/appmaster/runit/service/flink/run]: I follow the pattern there.

File: container/appmaster/runit/service/flink/run (hunk 1 above)
Comment (skonto) [inline container/appmaster/runit/service/flink/run]: will fix thnx.

Comment (skonto) [inline container/appmaster/runit/service/flink/run]: :+1:

**Comment by skonto** [comment]
@EronWright I updated the PR. Will test and report back.

File: container/appmaster/runit/service/flink/run (hunk 2 above)
Comment (EronWright) [inline container/appmaster/runit/service/flink/run]: Have a look at the `add_if_non_empty` function, it is basically a wrapper for appending to `FLINK_JAVA_OPTS`.   For the sake of uniformity, please use it to set the `high-availability` flag.

I have a minor concern about the sheer number of settings being configured as dynamic properties, as opposed to being configured in `flink-conf.yaml`.  Perhaps some of these HA settings can be set statically.

Comment (skonto) [inline container/appmaster/runit/service/flink/run]: Yeah sure for the first option I missed that. By the way it wont work for every property, there is one that causes problems:
2017-08-02 00:00:20,277 INFO  org.apache.flink.mesos.runtime.clusterframework.MesosApplicationMasterRunner  -     -Dhigh-availability.job.delay="10
2017-08-02 00:00:20,277 INFO  org.apache.flink.mesos.runtime.clusterframework.MesosApplicationMasterRunner  -     s"

The property Dhigh-availability.job.delay="10 s" is not preserved when passed.
In on order to fix this and keep a global variable you need to pass each arg quoted or use eval. The first option is not handy as we have many variables and the second one is not a good choice since it opens security problems if we don't sanitize every input. It gets ugly so I will try HA options in the file. Meanwhile, I am not sure it is going to work since the config.sh script removes quotes and this property needs it. I will test.
Btw I dont think it matters what goes as to the file or as dynamic properties from what I have seen in the code.

**Comment by skonto** [comment]
@EronWright fixed. Here is the guide how to run an example with the default settings:
https://gist.github.com/skonto/c8404d43142070dca860dda431459887

**Comment by skonto** [comment]
https://github.com/dcos/examples/pull/196

**Review by joerg84 — APPROVED** [review]

**Comment by joerg84** [comment]
@EronWright is this good to go? I would like include it in the Flink 1.3.2 release of the package.

**Review by tillrohrmann — APPROVED** [review]
LGTM. +1 for merging.

**Comment by skonto** [comment]
@EronWright @joerg84 how do you want me to proceed?

**Comment by skonto** [comment]
@EronWright @joerg84  gentle ping...

**Comment by MQasimSarfraz** [comment]
Are we planning to merge that?  this will be useful for our team as well?

**Comment by skonto** [comment]
@joerg84 gentle ping.

**Comment by skonto** [comment]
@joerg84 gentle ping...

**Comment by MQasimSarfraz** [comment]
Folks,

- I left my old job.
- Got married.
- Moved to new country.
- Settled in the new job.

but this PR is still in review! :disappointed:

**Comment by skonto** [comment]
closing this.
